### ISSUE TITLE
General question re: development - consistency of commands with matplotlib

### ISSUE THREAD

**Issue by catubc** [kushalkolar/fastplotlib#122]
Hi Kushal
This library looks pretty amazing. As a fellow [ca] imaging neuroscientist I greatly welcome this tool.

However, like so many out there it would be truly amazing if we can mostly drop in / replace exiting matplotlib routines with fastplotlib (with some tweaks of course).  How close will the nomenclature and syntax of fastplotlib be to matplotlib?  

Thanks a lot and looking forward to more developments.
catubc

**Comment by kushalkolar** [comment]
Hi, thanks for you comments. While I can understand the expectation to create a matplotlib-like API, we've come to the realization over the past ~year that the `pygfx` rendering engine allows us to create a much nicer, more elegant, and more expressive API than `matplotlib`. 

With that said, we also think that `fastplotlib` should be easy to pick up, have you tried out the examples? Start with [`simple.ipynb`](https://github.com/kushalkolar/fastplotlib/blob/master/examples/simple.ipynb). We've tried to make it intuitive which is one place where `matplotlib` often falls short.

For example Graphic features such as line colors can be sliced, if we start with:
```python
cosine_graphic = plot.add_line(data=<cosine_data>, thickness=12, cmap="autumn")
```

![before](https://user-images.githubusercontent.com/9403332/209256187-aab7b24d-c282-4a3a-b521-c4f4f6ececcf.png)

```python
cosine_graphic.colors[:15] = "magenta"
cosine_graphic.colors[70:] = "blue"
cosine_graphic.colors[60] = "w"
cosine_graphic.colors[25:55:3] = "cyan"
cosine_graphic.colors[75:90, 0] = np.linspace(0, 1, 15)
```

![after](https://user-images.githubusercontent.com/9403332/209258894-d7e5dbfe-3791-4319-907a-68526a5c882c.png)

Whereas with [`matplotlib` it's not so trivial](https://matplotlib.org/stable/gallery/lines_bars_and_markers/multicolored_line.html).  In addition, by leveraging `pygfx`'s capabilities, the GPU buffer is updated only at indices where it has changed (this is all transparent to the user and occurs when they assign values to the slice as shown above). This helps with performance and allows interactivity & modifications to data even with _millions_ of points in multiple subplots. See: https://www.youtube.com/watch?v=j_gwi-Wf1Ao 

With `fastplotlib`, we've tried to take inspiration from the best parts of `matplotlib`, `pyqtgraph`, and `bokeh` while also providing nicer ways to do things because of `pygfx`'s capabilities. Furthermore, `fastplotlib` is focused on fast interactive visualization, and `matplotlib` isn't really meant for interactive visualization so I think it would be a mistake to heavily base our API around `matplotlib` - different libraries exist for different reasons and I don't think one single library can cater to all the various visualization needs across so many different fields.

While this might not be the answer you were hoping for, there are always times when new tools or technologies come around and you wish you could just drop it in to existing libraries or workflows; but it's not always the best option (or possible) to make them backwards compatible. I worked with `pyqtgraph` for 5 years before starting `fastplotlib`. Anyways, it's up to the users on whether they want to revamp existing codebases with `fastplotlib` or just use it for new projects.

**Comment by catubc** [comment]
Hi Kushal
Thank you so much for the detailed response and effort to address my question. 
I agree that pure replication of matplotlib backend is not ideal.  However, for fast adoption and many projects, syntax similarity would help a lot.

I will do my best to try fastplotlib in the next while - my main worry is that some of the functionality I need (I've written a realtime brain-machine-interface GUI) will not be portable/implementable. So I might get stuck after dozens of hours of porting. Any advice there?

Thanks so much,
catubc

**Comment by kushalkolar** [comment]
I can't say much that's of use since I don't know your codebase. You can take a look at the current fastplotlib API and judge for yourself if you want to port it or not.

**Comment by kushalkolar** [comment]
I will close this, feel free to open another issue if you have API ideas after going through some example.
